Enhance TripPolicy for more Private Lines, filter for PassengerCategory, Access Modes

ue71603 opened the pull request:
With this PR it is possible to allow a trip planner to include specialised private LINE (e.g. car pooling, school busses, company shuttles) into the trip planning. 
It also allows to define necessary ACCESS MODE (e.g. that you are in a car, which restricts search to road, ferry, carTransportRail or that you can access a dragLift, because you have skis with you.
The third part makes sure, that selected by the PassengerCategory some additional things might get "unlocked". E.g. in Delemont seniors have a different service offered to them by the demand responsive service: https://www.postauto.ch/de/fahrplan-und-netz/publicar/delemont

skinkie commented on OJP/OJP_Trips.xsd:
```diff
@@ -1,8 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<xs:schema xmlns="http://www.vdv.de/ojp" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:siri="http://www.siri.org.uk/siri" targetNamespace="http://www.vdv.de/ojp" elementFormDefault="qualified" attributeFormDefault="unqualified">
+<xs:schema xmlns="http://www.vdv.de/ojp" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:siri="http://www.siri.org.uk/siri" xmlns:ifopt="http://www.ifopt.org.uk/ifopt" targetNamespace="http://www.vdv.de/ojp" elementFormDefault="qualified" attributeFormDefault="unqualified">
```
Is there a reason you include IFOPT here?

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,21 @@
 			</xs:element>
+			<xs:element name="PassengerCategory" type="PassengerCategoryEnumeration" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases (e.g. demand responsive, special areas are covered for seniors).</xs:documentation>
```
```suggestion
					<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases additional offerings may be available (e.g., demand responsive service with certain areas being reserved for seniors).</xs:documentation>
```

trurlurl reviewed (changes requested):
Sorry for slowing down the approval...

ue71603 commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,21 @@
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Access modes needed/supported for the trip. E.g. if one goes by car, then only ferry and carTransportRail may be possible for public transport. For dragLift skis or a snowboard may be necessary. When nothing is provided in the time table, then foot and bicycle should be assumed by the trip planner.</xs:documentation>
```
```suggestion
					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service (thus no need for needed and supported. E.g. if one goes by car, then only ferry and carTransportRail may be possible for public transport. If the traveler has skis or a snowboard with him, then dragLifts can show up in the results to a search. When nothing is provided in the time table, then foot and bicycle should be assumed by the trip planner.</xs:documentation>
```

ue71603 commented:
The use cases are:

•	Alice wants to go to the mountain top restaurant and uses a snowboard.  As the mountain top restaurant during winter is accessible by a dragLift, this should show in the trip result, because Alice sent (foot ski) in the new AccessModeList parameter.
•	Bob goes from Bern to his holiday hotel in Crans Montana. He will use the carTransportRail between Kandersteg and Goppenstein. He informs himself about the waiting time and specifies his desired arrival time in Goppenstein, this results in a rail trip between Kandersteg and Goppenstein and he expands this with two trip requests (Bern – Kandersteg; Goppenstein – Crans Montana).
•	Charlie wants to go to Crete. She is going with the motorcycle and needs to use a ferry that is willing to transport her and the vehicle.
•	“Auto im Reisezug” we don’t have that really, but I guess this is also a use case.
•	In future we will have additional use cases for Cargo transport and there it might be relevant).

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,21 @@
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service (thus no need for needed and supported. E.g. if one goes by car, then only ferry and carTransportRail may be ...
```
So the planner would only return PT routes where I can take all of my *AccessModes* with me, for the complete route.  With *AccessMode* ski I would get a route from Goppenstein to Kandersteg by train and further up using a drag lift to some peak. With *AccessModes* ski and car I wouldn't get any result for the same departure and destination; but I would be able to get a result for Goppenstein to Kandersteg.

Is should *wheelchair* also be an *AccessMode*?

ue71603 replied:
For BeTax yes we should

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,6 +107,18 @@
+			<xs:documentation>List of Access Modes.</xs:documentation>
+		</xs:annotation>
+		<xs:sequence>
+			<xs:element name="AccessMode" type="ifopt:AccessModesEnumeration" minOccurs="0">
```
Is a *maxOccurs="unbounded"* missing here? And couln't the term *List* be misleading, since it wouldn't be a real list where multiple items are possible between the element tags (*<<AccessModeList>>ski car bicycle<</AccessModeList>>*)?

ue71603 replied:
yes, it should.

ue71603 replied:
done

ue71603 commented on OJP/OJP_ModesSupport.xsd:
```diff
@@ -30,10 +30,13 @@
+			<xs:documentation>[a category of MODE in TMv6] MODEs offered by private individuals (or companies or closed communities).</xs:documentation>
 		</xs:annotation>
 		<xs:restriction base="xs:string">
 			<xs:enumeration value="carPooling"/>
```
Least: any one.

ue71603 replied:
done

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,6 +107,18 @@
+			<xs:documentation>List of Access Modes.</xs:documentation>
+		</xs:annotation>
+		<xs:sequence>
+			<xs:element name="AccessMode" type="ifopt:AccessModesEnumeration" minOccurs="0">
```
```suggestion
			<xs:element name="AccessMode" type="ifopt:AccessModesEnumeration" minOccurs="0" maxOccurs="unbounded">
```

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -122,6 +135,21 @@
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service. E.g., if one goes by car, then only ferry and carTransportRail may be possible for public transport. If the ...
```
Shouldn't the filtering behavior based on *AccessModeList* be more clearly stated here? If Goppenstein-Kandersteg did support car but not bicycle, what would happen if I had both of them in the *AccessfModeList*? Would the journey be excluded (not possible to transport all of them) or included? What should be the behavior if I want to go from Goppenstein to the top of a drag lift above Kandersteg and I have car and ski in the *AccessModeList*? Should the algorithm be so intelligent to propose Goppenstein-Kandersteg by train, followed by a car drive to the drag lift, followed by the drag lift (transporting the skis but not the car)? Or has every leg to be queried separately? (But this shouldn't be done according to the implementors.)

ue71603 commented on OJP/OJP_Trips.xsd:
```diff
@@ -122,6 +135,21 @@
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service. E.g., if one goes by car, then only ferry and carTransportRail may be possible for public transport. If the ...
```
I will add a better description.  If one fulfills one of the criteria, one is allowed to board.  
Malte: car drive  and the car follows afterwards.  Examples handling.

ue71603 replied:
customer [foot motorcycle ski]

AccessMode can transport AccessMoe
motorcycle   -> ski foot 
car   -> ski foot bicycle

road [*] - carTransportRail [car motorcycle] - dragLift [ski] -

ue71603 replied:
André: One gives the list. For each leg you get for each item, RequiredList

ue71603 commented:
Tim: Another use case. If one uses more than journey planner and link the answers. One needs to pass the information between journey planners. The national service planner was based on regional service planners. So the previous context needs to be transfered.
Matthias: The main behaviour of trip planners does not change. 
--> SBB also prepares a detailed example to be included.

skinkie commented on OJP/OJP_ModesSupport.xsd:
```diff
@@ -30,10 +30,14 @@
+			<xs:documentation>[a category of MODE in TMv6] MODEs offered by private individuals (or companies or closed communities).</xs:documentation>
 		</xs:annotation>
 		<xs:restriction base="xs:string">
-			<xs:enumeration value="carPooling"/>
```
Any reason you have removed carPooling all together?

ue71603 replied:
I think we discussed that vehiclePooling covers all. You disagree?

skinkie replied:
Ok, fine for me then.

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,21 @@
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service. E.g., if one goes by car, then only ferry and carTransportRail may be possible for public transport. If the ...
```
```suggestion
					<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service. E.g., if one goes by car, then only ferry and carTransportRail may be possible for public transport. If the traveller has skis or a snowboard with him, then dragLifts can show up in the results to a search. When nothing is provided in the time table, then foot and bicycle should be assumed by the trip planner. The resulting trips satisfy the following conditions: (a) the PT legs of the trip are possible for all access modes in the list, (b) the non-PT legs are possible by foot or any of the access modes in the list. In the end the provision of the list does not guarantee valid results in all circumstances. E.g. the traveller may have left the car, went with the skis up the hill in a dragLift and can't access the car on the other side of the mountain for the next leg with carTransportRail. So the enquirer always needs to study the trip results to see what AccessMode and/or are provided there. The implementation might also have an internal mapping which AccessMode can "carry" another one (e.g. a car can carry skis).</xs:documentation>
```

ue71603 replied:
<xs:documentation>Access modes that are available to the traveller for the trip. The requester can tell the system what he might use as an access mode (e.g., car, motorcycle, ski) and it's the system's responsibility to know which access modes are supported or required by each service. E.g., if one goes by car, only then car ferry and carTransportRail may be possible for public transport. If the traveller has skis or a snowboard with him, then dragLifts can show up in the results to a search. When nothing is provided in the time table, then foot and bicycle should be assumed by the trip planner. The resulting trips satisfy the following conditions: (a) the PT legs of the trip are possible for all access modes in the list, (b) the non-PT legs are possible by foot or any of the access modes in the list. In the end the provision of the list does not guarantee valid results in all circumstances. E.g. the traveller may have left the car, went with the skis up the hill in a dragLift and can't access the car on the other side of the mountain for the next leg with carTransportRail. So the enquirer always needs to study the trip results to see what AccessMode and/or are provided there. The implementation might also have an internal mapping which AccessMode can "carry" another one (e.g. a car can carry skis).</xs:documentation>

Request AccessMode: AccessMode the traveler can offer at the origin 

 [foot car ski] 
 
 examples:
 * (walk - )car - carTransportRail - car - walk - dragLift - walk
											-> losing car      -> [foot ski]
 *  (walk -) dragLift - walk - carTransportRail - walk  (nonsense)
 
Delivery AccessMode: AccessMode that are supported by a SITE.


BLS: bicycle motorcycle car (truck)
MGB:  (foot) (motorcycle) car (truck)
RhB:   motorcycle car
HuPac:  truck

Mapping
ski -> car
foot -> car
bicycle -> car

Norman: The need in the distributed manner. More generalized about context to be provided in the DJP.

ue71603 commented:
@skinkie  @normanoffel @herlitze @DRubliMentz 
I moved IncludeAlternativeOptions to the TripPolicyGroup. I reorganised the TripPolicyGroup a little bit.
I added MainTravelMode. When this is set to "car" then railTransport etc is calculated and IV. Otherwise carTransportRail is excluded.  Does this sound reasonable. This also does a little bit what the IVProfile does in our current implementation.
Or do you prefer something else?

skinkie commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,31 @@
+			</xs:element>
+			<xs:element name="MainTravelMode" type="ifopt:AccessModesEnumeration" default="foot" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>The MainTravelMode defines whether the trip planner works in PT mode or includes carTransportRail and the like. The only mode that activates the inclusion of this is "car". </xs:documentation>
```
```suggestion
					<xs:documentation>The MainTravelMode defines whether the trip planner works in PT mode or includes carTransportRail and the like. The only mode that activates the inclusion of this is "car".</xs:documentation>
```

ue71603 commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,31 @@
+					<xs:documentation>The MainTravelMode defines whether the trip planner works in PT mode or includes carTransportRail and the like. The only mode that activates the inclusion of this is "car". </xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="AccessModeList" type="AccessModesListOfEnumerations" minOccurs="0">
```
Removed AccessModeList in the request. Added IncludeTouristicOffers instead.

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -117,6 +130,31 @@
+				<xs:annotation>
+					<xs:documentation>The MainTravelMode defines whether the trip planner works in PT mode or includes carTransportRail and the like. The only mode that activates the inclusion of this is "car".</xs:documentation>
+				</xs:annotation>
+			</xs:element>
```
I'm sorry for again raising questions because I still don't sufficiently understand the aim.

**_MainTravelMode_** car (would also work for motorcycle). What is expected from the routing algorithm?
	1. Variant 1: With the exception of the initial and final access legs, the whole trip is accessible for the car. 
	2. Variant 2: A substantial part of the trip, consisting of a connected sequence of legs, is accessible for the car. 
	3. Variant 3: The planning is done like for a normal passenger, but with additional services unlocked that might allow for additional or faster solutions.

In variants 1 and 2, the routing algorithm searches for trips involving at least one PT leg and minimising the distance covered by the continuous legs. (Or is that too complicated because of the combination of PT and continuous planning?)
In variant 3, MainTravelMode car would just unlock additional services, the algorithm would not care whether the trip is feasible or somehow optimal for an automobilist.

Both 1 and 2 would be able to find the car transport rail Goppenstein-Kandersteg, or a ferry allowing to shorten driving distance.
Variant 3 might also find them, but only "by chance", if they appear in a pure PT / time table search.
Variant 1 would be able to find a nightly car transport rail service Basel-Hamburg allowing to shorten driving distance. 
Variant 2 might limit the nightly car transport rail to Basel-Freiburg and propose to continue by a much faster ICE, leaving the car in Freiburg. This I would consider a strong argument against variant 2. 
Variant 3 would be able to consider the nightly car transport service, but probably would abandon it as it is very slow.
Variants 2 and 3, but not 1, could find a trip involving car transport rail Goppenstein-Kandersteg and a drag lift to a mountain top given MainTransportMode car and IncludeTouristicOffers.


**_IncludeTouristicOffers_** (and access modes as proposed previously)  -  unlocks additional services

Without a MainTransportMode like car or motorcycle ->  a pure PT search having access to unlocked additional services (like variant 3)

If access mode car still existed, this would allow for finding a trip involving car transport rail Goppenstein-Kandersteg and a drag lift to a mountain top given access mode car and IncludeTouristicOffers. Thus circumventing the limitation of variant 1 mentioned above.


Additional considerations: 
Transfer legs: variant 1 would not include any transfer legs because _paths between PT legs have to be continuous legs_ (car)
Transfer legs without a main mode like car or motorcycle: should the algorithm try to replace transfer legs with continuous legs if an adjacent PT leg is accessible by car? 
How about bicycles? 

**Preliminary conclusion:**

The **most useful solution** seems to me: variant 1 for MainTransportMode car and motorcycle, And to keep access modes. For me variant 1 feels like invoking an optimisation method aiming to propose attractive car transport offers.

The **simplest solution** seems to me, but of very limited use: variant 3, but for clarity I would propose to rename _MainTransportMode_ to _IncludeVehicleTransportOffers_ (for car, motorcycle, bicycle).

trurlurl replied:
The choice construct to separate behaviors (1) and (3) seems like a good solution to me.

I think keeping MainTransportMode is better if/as the intention is variant (1).

I think IncludeTouristicOffers is also a good solution as it allows to simply unlock a bunch of services and it is the responsability ot the user to determine whether he can use it (necessary equipment, consistency of trip). But how is he informed about that? There is the _AccessModeList_ for places, but I don't see how this could tell something about trip legs.

How about motorcycles? If they're used to be considered a "kind of car", _car_ may be sufficient in MainTransportMode to activate the behavior (1), otherwise I think _motorcycle_ should also be valid/activating.

How about bicycles? I think the best behavior would be: if the user selects _BikeTransport_ (TripMobilityFilter) then the filter should  not only restrict the search, but unlock services for which bicycle is an access mode.

ue71603 commented:
@trurlurl and now?

trurlurl reviewed (dismissed):
Looks good to me!

skinkie commented on OJP/OJP_Common.xsd:
```diff
@@ -194,10 +194,20 @@
 		</xs:annotation>
 		<xs:sequence>
-			<xs:element ref="JourneyRef"/>
+			<xs:element ref="JourneyRef" minOccurs="0"/>
```
```suggestion
			<xs:element ref="JourneyRef"/>
```

ue71603 replied:
```suggestion
			<xs:element ref="JourneyRef"/>
```

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+		<xs:sequence>
+			<xs:group ref="BaseTripPolicyGroup"/>
+			<xs:choice>
+				<xs:element name="MotorizedMainTravelMode" type="xs:boolean" minOccurs="0">
```
```suggestion
				<xs:element name="MotorisedMainTravelMode" type="xs:boolean" minOccurs="0">
```

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+			<xs:choice>
+				<xs:element name="MotorizedMainTravelMode" type="xs:boolean" minOccurs="0">
+					<xs:annotation>
+						<xs:documentation>The MotorizedMainTravelMode defines whether the trip planner works in PT mode (or pure IT mode) or includes carTransportRail and the like. If set to true (e.g. because one travels by car, truck or motorcycle), then carTransportRail, ferry are used together with roads.</xs:documentation>
```
```suggestion
						<xs:documentation>The MotorisedMainTravelMode defines whether the trip planner works in PT mode (or pure IT mode) or includes carTransportRail and the like. If set to true (e.g. because one travels by car, truck or motorcycle), then carTransportRail, ferry are used together with roads.</xs:documentation>
```

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+					</xs:element>
+					<xs:element name="IncludeRestrictedLines" type="xs:boolean" minOccurs="0" default="false">
+						<xs:annotation>
+							<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an AccessMode ski. If this flag is set, then existing restricted lines are considered by the router. The results are marked as restricted in the ServiceSection. The restriction can be seen from: ACCESS MODE or Pr ...
```
```suggestion
							<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an AccessMode ski. Lines with ACCESS MODE bicycle will be included as well. If this flag is set, then existing restricted lines are considered by the router. The results are marked as restricted in the ServiceSection. The restriction can be seen from: ACCESS MODE or PrivatMode. The usage could also be detailed with Attribute elements.</xs:documentation>
```

trurlurl replied:
Please have a look at the bicycle proposal; if it should be the other way round, I think it would be good to state that, too.

ue71603 replied:
@trurlurl  bicycles are available by default (as you can use them on most publi transport)

ue71603 replied:
Im Fall, den du erwähnst («bicycles are available by default») ist das bicycle nicht access mode, man braucht es ja nicht um mitfahren zu dürfen. Beim Beispiel Goppenstein-Kandersteg ist es (oder könnte es zumindest sein) ein access mode, neben car und motorcycle.

Mein Punkt hier wäre: ich fände es gut, klar zu definieren, wie das bicycle behandelt werden soll, damit man als requester weiss, ob bicycle als access mode mitgeteilt wird oder nicht. 

Für mich ist die Sache eigentlich auch bei car, motorcycle, truck nicht so ganz eindeutig. Nach der annotation und der Tatsache, dass es bereits MotorisedMainTravelMode gibt, würde ich eher nicht damit rechnen, dass car- oder motorcycle-Strecken freigeschaltet werden. Aber der Name «IncludeRestrictedLines»  alleine würde vermuten lassen, dass jegliche restricted lines berücksichtigt werden. 

Ich glaube, dass es die bessere Lösung wäre, wenn bicycle berücksichtig wird bei IncludeRestrictedLines, und nicht nur wenn BikeTransport angewählt ist. Z.B. bei dragLifts mit access mode bike, da kommt einem erst durch das Angebot die Idee, ein Velo mitzunehmen.

Wenn der Annotationstext angepasst wird, sollte auch die annotation zu IncludeRestrictedLines in OJP_StopEvents.xsd angepasst werden.

@trurlurl will use your suggest. pls do the change directly in all places, where it needs to be done.

trurlurl commented on OJP/OJP_StopEvents.xsd:
```diff
@@ -56,6 +56,11 @@
 				</xs:annotation>
 			</xs:element>
+			<xs:element name="IncludeRestrictedLines" minOccurs="0">
+				<xs:annotation>
```
Comment would have to be aligned if changed at the other place. Just to make sure because I don't understand how exactly it will work - Will putting the access mode on the place where the leg starts be working in all circumstances to indicate the restriction on this line?

ue71603 replied:
@trurlurl no. We have multiple scenarios:
* the ACCESS MODE restricts the LINE (e.g. dragLift)
* The needed qualification restricts the usage of the LINE: e.g. you need to be a pupil going to a given schoole, you need to stay at a given hotel, work for a given company.
But in the request we don't differenciate: They will just all show up.

trurlurl replied:
@ue71603 My concern was just that ACCESS MODE restricts the LINE only indirectly (as it is attached to a PLACE, not the line). And I'm not sure if the return structures from StopEventRequest and TripRequest always contain the initial PLACE.

skinkie commented on OJP/OJP_Trips.xsd:
```diff
@@ -1581,11 +1613,7 @@ Can only be used when both participants recognise the same schedule version. If
 					<xs:documentation>System reference to use for the refinement. If not specified the origin systems of each leg are used for the refinement.</xs:documentation>
 				</xs:annotation>
 			</xs:element>
-			<xs:element name="UseRealtimeData" type="UseRealtimeDataEnumeration" default="full" minOccurs="0">
```
Question: where did UseRealtimeData move to?

herlitze replied:
I think it was removed by mistake

herlitze commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+							<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an ACCESS MODE ski. Lines with ACCESS MODE bicycle will be included as well. If this flag is set, then existing restricted lines are considered by the router. The results are marked as restricted in the ServiceGr ...
+						</xs:annotation>
+					</xs:element>
+					<xs:element name="PassengerCategory" type="PassengerCategoryEnumeration" minOccurs="0" maxOccurs="unbounded">
```
@ue71603 What are the expected results, if I pass several categories? Would these service for seniors be included, if I IncludeRestrictedLines?

ue71603 replied:
IncludeRestrictedLines trumps PassengerCategory. For PassengerCategory I suggest that it is the union. If a Service is available lfor one of the PassengerCategories it is shown.

ue71603 replied:
next version. @trurlurl  can you check?

ue71603 commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -143,6 +143,11 @@
 			</xs:element>
+			<xs:element name="Restricted" type="xs:boolean" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>This flag is set, when the Service has restricted usage. E.g. a certain ACCESS MODE is required (e.g. dragLift) or if the LINE is provided only to special sets of passengers (e.g. school buses, hotel shuttles) as defined in PrivateMode.</xs:documentation>
```
```suggestion
					<xs:documentation>This flag is set, when the Service has restricted usage. E.g. a certain ACCESS MODE is required (e.g. dragLift) or if the LINE is provided only to special sets of passengers (e.g. school buses, hotel shuttles) as defined in PrivateMode. Restricted trumps e.g. PassengerCategory. If Restricted is set to true all Service are shown.</xs:documentation>
```

ue71603 commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+					</xs:element>
+					<xs:element name="PassengerCategory" type="PassengerCategoryEnumeration" minOccurs="0" maxOccurs="unbounded">
+						<xs:annotation>
+							<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases additional offerings may be available (e.g., demand responsive service with certain areas being reserved for seniors).</xs:documentation>
```
```suggestion
							<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases additional offerings may be available (e.g., demand responsive service with certain areas being reserved for seniors). If multiple PassengerCategories are provided, then Services are shown when the are available for at least one PassengerCategory.</xs:documentation>
```

ue71603 commented:
@herlitze  When do you have time to have a look at the 2 PR?

trurlurl reviewed (changes requested):
Open question regarding PassengerCategory.

ue71603 commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+					</xs:element>
+					<xs:element name="PassengerCategory" type="PassengerCategoryEnumeration" minOccurs="0" maxOccurs="unbounded">
+						<xs:annotation>
+							<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases additional offerings may be available (e.g., demand responsive service with certain areas being reserved for seniors). If multiple PassengerCategories are provided, then Services are shown when the are available for at least one PassengerCategory.</xs:documentation>
```
```suggestion
							<xs:documentation>Sequence of all passenger categories, for which this search shall be conducted. In rare cases additional offerings may be available (e.g., demand responsive service with certain areas being reserved for seniors). If multiple PassengerCategories are provided, then Services are shown when they are available for at least one PassengerCategory.</xs:documentation>
```

ue71603 commented:
> Open question regarding PassengerCategory.

@trurlurl : What is still open Dog und Bicycle are in the PassengercategoryEnumeration. So it need to "work". do you want a textual addition.

> I haven't found where the restrictions applying to a _passenger category could be seen. E.g., the Delémont case. Would it have to be modeled as a PrivateMode?

You want an update on the example?

trurlurl commented:
> > Open question regarding PassengerCategory.
> 
> @trurlurl : What is still open Dog und Bicycle are in the PassengercategoryEnumeration. So it need to "work". do you want a textual addition.
> 
> > I haven't found where the restrictions applying to a _passenger category could be seen. E.g., the Delémont case. Would it have to be modeled as a PrivateMode?
> 
> You want an update on the example?

@ue71603 Yes, I think having a "dogs not allowed" and "only for seniors" example would be helpful to show where a calling system has to collect such information in order to present it to the enquirer.

ue71603 commented:
@trurlurl : Happy now?

sgrossberndt commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -143,6 +143,11 @@
 			</xs:element>
+			<xs:element name="Restricted" type="xs:boolean" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>This flag is set, when the Service has restricted usage. E.g. a certain ACCESS MODE is required (e.g. dragLift) or if the LINE is provided only to special sets of passengers (e.g. school buses, hotel shuttles) as defined in PrivateMode. Restricted trumps e.g. PassengerCategory. If Restricted is set to true all Service are shown.</xs:documentation>
```
Proposal:
This flag is set if the service can only be used in a restricted way. For example, a specific ACCESS MODE is required (e.g. dragLift) or the LINE is only made available to certain passenger groups (e.g. school buses, hotel shuttles), as defined in PrivateMode. Restricted trumps PassengerCategory, for example. If Restricted is set to true, all services are displayed.

ue71603 replied:
changed

sgrossberndt commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -216,7 +221,7 @@
 	<xs:group name="PrivateServiceGroup">
 		<xs:annotation>
-			<xs:documentation>specific service operated by a Private Mode</xs:documentation>
+			<xs:documentation>specific service operated by a Private Mode. May also be a school bus, a company shuttle or some other transport opportunity for a (normally) closed community.</xs:documentation>
```
Proposal:
Specialized service operated by a Private Mode, e.g. a school bus, corporate shuttle, or other transportation for a (usually) closed community.

ue71603 replied:
included

sgrossberndt commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -225,6 +230,16 @@
 					<xs:documentation>[a category of MODE in TMv6] MODEs offered by private individuals </xs:documentation>
```
trailing space in excess

ue71603 replied:
changed.

sgrossberndt commented on OJP/OJP_StopEvents.xsd:
```diff
@@ -56,6 +56,11 @@
 			</xs:element>
+			<xs:element name="IncludeRestrictedLines" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an ACCESS MODE ski. Lines with ACCESS MODE bicycle will be included as well. If this flag is set, then existing restricted lines are considered by the router. The results are marked as restricted in the ServiceGrou ...
```
Proposal:
There may be lines that are subject to special restrictions and are not generally open to the public. E.g. school buses, company shuttles. dragLifts must have an ACCESS MODE ski. Lines with ACCESS MODE bicycle are also considered. If this flag is set, existing restricted lines are considered by the router. The results are marked as restricted in the ServiceGroup. The restriction can be recognized by: ACCESS MODE or PrivateMode. The usage can also be detailed with attribute elements.

ue71603 replied:
changed

sgrossberndt reviewed (dismissed):
Are all of these changes actually related to each other in a way they have to be merged in one pull request? It isn't really easy to understand.

ue71603 commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -143,6 +143,11 @@
 			</xs:element>
+			<xs:element name="Restricted" type="xs:boolean" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>This flag is set, when the Service has restricted usage. E.g. a certain ACCESS MODE is required (e.g. dragLift) or if the LINE is provided only to special sets of passengers (e.g. school buses, hotel shuttles) as defined in PrivateMode. Restricted trumps e.g. PassengerCategory. If Restricted is set to true all Service are shown.</xs:documentation>
```
```suggestion
					<xs:documentation>This flag is set if the service can only be used in a restricted way. For example, a specific ACCESS MODE is required (e.g. dragLift) or the LINE is only made available to certain passenger groups (e.g. school buses, hotel shuttles), as defined in PrivateMode. Restricted trumps PassengerCategory, for example. If Restricted is set to true, all services are displayed.</xs:documentation>
```

ue71603 commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -216,7 +221,7 @@
 	<xs:group name="PrivateServiceGroup">
 		<xs:annotation>
-			<xs:documentation>specific service operated by a Private Mode</xs:documentation>
+			<xs:documentation>specific service operated by a Private Mode. May also be a school bus, a company shuttle or some other transport opportunity for a (normally) closed community.</xs:documentation>
```
```suggestion
			<xs:documentation>Specialized service operated by a Private Mode, e.g. a school bus, corporate shuttle, or other transportation for a (usually) closed community.</xs:documentation>
```

ue71603 commented on examples/functions/trip/TR_07_request_Delemont_senior.xml:
```diff
@@ -0,0 +1,42 @@
+					</PlaceRef>
+				</Destination>
+				<Params>
+					<PassengerCategory>Senior</PassengerCategory>
```
@trurlurl  If the PassengerCategory is set, there is no need for the IncludeRestrictedLines. It shows up. I will add an additional example showing it and a "private service".

trurlurl replied:
Aha - to make this more clear, see the proposition on line 147.

ue71603 commented on examples/functions/trip/TR_08_request_dog.xml:
```diff
@@ -0,0 +1,44 @@
+				</Destination>
+				<Params>
+					<PassengerCategory>Senior</PassengerCategory>
+					<PassengerCategory>Dog</PassengerCategory>
```
Will include this in the new example.

ue71603 commented on examples/functions/trip/TR_08_response_dog.xml:
```diff
@@ -0,0 +1,330 @@
+			<siri:ErrorCondition>
+				<OJPError>
+					<TripProblemType>TRIP_RESTRICTEDRESULTSDROPPED</TripProblemType>
+				</OJPError>
```
I think most people will not implement this. I think this can be used, when the service knows it dropped things because they are restricted.  In this case the Publicar was droopped, because: (a) no Senior (b) no IncludeRestrictedLines.

trurlurl replied:
OK.

ue71603 replied:
I improved the example some more.

trurlurl replied:
I think the PubliCar shouldn't be dropped, since ,,, "If multiple PassengerCategories are provided, then Services are shown when they are available for at least one PassengerCategory." 

And in that case the error condition wouldn't appear and should be shown in another example.

ue71603 replied:
now only dog, even when this doesn't make completly sense.

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+					</xs:element>
+					<xs:element name="IncludeRestrictedLines" type="xs:boolean" minOccurs="0" default="false">
+						<xs:annotation>
+							<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an ACCESS MODE ski. Lines with ACCESS MODE bicycle will be included as well. If this flag is set, then existing restricted lines are considered by the router. The results are marked as restricted in the ServiceGr ...
```
```suggestion
							<xs:documentation>There might exist lines that have special restrictions and are not generally available to the public. E.g. school buses, company shuttles. dragLifts need to have an ACCESS MODE ski. Lines with ACCESS MODE bicycle will be included as well. If this flag is set, then existing restricted lines are considered by the router, irrespective of the selected passenger categories or the MotorisedMainTravelMode. The results are marked as restricted in the ServiceGroup. The restriction can be seen from: ACCESS MODE or PrivateMode. The usage could also be detailed with Attribute elements.</xs:documentation>
```

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
+							<xs:documentation>Whether alternative options should be presented as well. Mainly important for dominated journeys or in the case of ContinuousLegs the second-best route. Should be optimised for the user expectance (see. e.g https://theses.hal.science/tel-01848737). However, what the alternative options are may vary widely depending on the optimisation methods and filters.</xs:documentatio ...
+						</xs:annotation>
+					</xs:element>
+					<xs:element name="IncludeRestrictedLines" type="xs:boolean" minOccurs="0" default="false">
```
Might `IncludeAllRestrictedLines` be a better name?

ue71603 replied:
changed

trurlurl commented on OJP/OJP_JourneySupport.xsd:
```diff
@@ -143,6 +143,16 @@
 			</xs:element>
+			<xs:element name="Restricted" type="xs:boolean" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>This flag is set if the service can only be used in a restricted way. For example, a specific ACCESS MODE is required (e.g. dragLift) or the LINE is only made available to certain passenger groups (e.g. school buses, hotel shuttles), as defined in PrivateMode. Restricted trumps PassengerCategory, for example. If Restricted is set to true, all services are displayed.</xs:doc ...
```
Are the last two sentences at the right place here? Or were they intended for `IncludeRestrictedLines`? That would match my comment on line 147.

trurlurl commented on OJP/OJP_Trips.xsd:
```diff
@@ -106,18 +107,54 @@
 		</xs:sequence>
 	</xs:group>
-	<xs:group name="TripPolicyGroup">
+	<xs:complexType name="AccessModesListOfEnumerations">
```
If that structure isn't used for trips any more, it could be moved to OJP_PlaceSupport.xsd

ue71603 replied:
done

ue71603 commented:
> Are all of these changes actually related to each other in a way they have to be merged in one pull request? It isn't really easy to understand.

They came out of the same meeting and discussion (on sharing and necessary extensions). I know they are difficult to understand. But breaking them apart would not make it better.
